### ISSUE TITLE
Big Vector problem

### ISSUE THREAD

**Issue by jolestar** [libra/libra#4377]
The Vector in Move is mapping to a Rust Vec at Runtime if the user creates a big vector,  such as tens of thousands or even millions of elements are in a vector, will cause performance issues, and use too much gas.

Currently, user's all resource serialize to an AccountBlob,  if a user has too many resources,  it will also cause this problem too.

Does Move have plans to fix this issue?

**Comment by bob-wilson** [comment]
I'm not sure what you're suggesting. Can you clarify? If the user creates a vector with millions of elements, then fundamentally it will be expensive. Are you just asking for some performance and/or space optimizations for vectors?

**Comment by aihuawu** [comment]
I am wondering how you compute gases for vectors.

I also have big vector for a big table. I understand the looping is expensive.
If i only call vec[i], one element access,
will this be an expensive operation? should I pay a lot of gas for it?

**Comment by jolestar** [comment]
Move VM loads the whole vector from storage into memory when user access vec[i] in Move. if this operator is cheep in gases, but it really is an expensive operation in VM Runtime, this will introduce a vulnerability to resource abuse. 

My suggestion is that Move's Vector is expressed as a reference in VM runtime, and is not needed to load the whole vector at runtime, only loads one an element from storage when the user actually accesses vec[i].  Ethereum saves the map and vector like this. This means that access_path should support expression accessing the resource field or index.

I see there an Access structure in libra's early AccessPath code. It was supposed to be used to express AccessPath for accessing a resource's field or vector index, but it was later cleaned up. I'd like to know what the thinking is on this?

```rust
pub enum Access {
    Field(Field),
    Index(u64),
}
```
Thanks.

**Comment by vgao1996** [comment]
Currently I don't think there's a way around: vectors are serialized as blobs and the VM reads them as a whole. (In fact, if I remember correctly, you must always load the whole top-level resource, which is even more terrifying for something like nested data structures.) As @jolestar mentioned, we cannot charge less than O(n) gas or otherwise it will be exploitable.

With that being said, I do believe it is important that we eventually support lazy & partial loading, as it will be a game changer for many applications. Those applications may be prohibitively expensive until we bring the cost of the following operations down to O(1):
- `borrow_global<vector<...>>`/`borrow_global<struct>`
- `borrow_field` from the values above
- push, pop, get, set
A simple example would be an ever-growing log, which may become too expensive to access very soon.

I have a vague intuition of how this could potentially be implemented, but I don't think it can come anytime soon as it requires
- Some major work to add lazy vectors and lazy structs to the runtime.
- Redesigned storage scheme for vector and structs.
  - A lot more complicated than what we have right now. We may even need to switch between two layouts (compact vs sparse) intelligently in order to prevent a major regression in storage efficiency.
- Some sort of garbage collection mechanism in storage. It may not be as simple as doing this eagerly for every writeset.

There are also some uncertainties I can think of:
- Gas metering may become complicated.
- Interaction with LCS. With the new storage scheme, a client needs to do more than just reading and deserializing a bunch of bytes from storage.

@dariorussi @tzakian @sblackshear @tnowacki @runtian-zhou what do you guys think? IMO this is challenging work we may explore in a long term.

**Comment by tnowacki** [comment]
While everything @vgao1996 said is true, I want to walk this back a bit. 

**You should always and forever assume `borrow_global` is going to be O(n) in gas cost**

In short, we will likely never add any sort of lazy loading for `borrow_global`. A few reasons why
- Currently, the on-chain storage API does not support partial reads. Its a more simple key-value store. So all of the bytes for a resource have to be copied over. 
- Even if we could support partial reads, adding lazy loading inside of the struct/vector would be an *incredibly* brittle optimization. It would really hamstring us for future space/layout optimizations we might want to do for internal representations of this data. As such it might not always kick in and then you occur a higher gas cost. Additionally, such an optimization might make adding new data structures very difficult. 
- I hesitate to even suggest the next point, but I will anyway. Just as a disclaimer: Assume we will not add this :p. Nonetheless, if we do ever add some sort of lazy-loading vector/struct, it would likely be in another type. Something like `ShardedVector` as another module with a bunch of utilities. Or some native type `sharded_vector`. 

As a whole, Move's storage layer is a very different beast than is found in Ethereum/Solidity. Due to the resource model, you can manage the splitting of data yourself into multiple resources! Imagine if you wanted to add another field to `LibraAccount` but were afraid of the data getting to large. Perhaps this new field is only read in on specific function. You could wrap this new data in another resource `AdditionalData` and as such you would not increase the cost of the normal code paths.
For example
```
module LibraAccount {
...
  resource struct AdditionalData { x: X }
  public fun add_new_data(account: &signer, x: X) {
    move_to(signer, AdditionalData { x }
  }
  public fun check_additonal_data(addr: address): bool { ... }
}
```
While the example doesn't do the best job highlighting it, none of the other code paths will ever load `AdditionalData`, thus preventing additional data from being loaded. 

In summary, Move's data model shies away from big monolithic store. Data that corresponds to an individual account  should be stored on that account; the data should not be aggregated into one centralized collection. If you *really* need a big storage or a big vector, you will have to do some work to shard it yourself, but it should be an indication that your design needs to be reworked. 

I'll tag team over to @sblackshear who has another example about what to do if you absolutely need some big centralized storage

**Comment by sblackshear** [comment]
"How do I create a big ol' map or vector" is a common question for folks familiar with programming with Ethereum, but it is rarely the right design pattern. To give a few examples:

Custom coin:
- Eth: use map of to int
- Move: define `Coin`(n) resource, give one to each address. No map needed

Auction:
- Eth: use map from address to bid
- Move: define `Bid`(n) resource, give it to bidder in exchange for `Coin`(n) resource. No map needed

DNS:
- Eth: use map from domain name to address
- Move: define `DomainName` resource with `next: address` and `prev: address` "pointers", hand one to each account that registers a `DomainName`. The logic for adding a domain name to an account can maintain a sorted doubly-linked list to check for duplicates. This one is more subtle than the previous ones; here is some example [code](https://github.com/libra/libra/blob/master/language/ir-testsuite/tests/sorted_linked_list/sorted-linked-list.mvir) (unfortunately old/in mvir).

Coming up with these design patterns is sometimes challenging, but interesting. Curious to hear the  use-cases you have in mind for big tables and whether they can be avoided with these patterns or another yet to be invented.

**Comment by jolestar** [comment]
@sblackshear  I agree with your design pattern example, but my question is not how to avoid large Vector in Move, but if the user saves too much data in Vector, how can we handle it? Blockchain is a public service, We either ensure that the user can't save too much data through cost or through hard-coded limitations in Move.

**Comment by sblackshear** [comment]
The answer we have designed/implemented is the first one you mention: if the user saves a lot of data in a vector, they will pay a high gas cost.
